### ISSUE TITLE
Moneris Credit Card Processing

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#393]
In GitLab by @tim.traylor on Jun 15, 2017, 07:13

Hi Guys,

The new external client, Select Call Center, uses Moneris to process credit card transactions.  We would like to have you add Moneris as a credit card gateway in SA Billing.

The API information can be found here: developer.moneris.com.  They have a very well documented API.

Please be sure to implement this gateway using the same mechanism and methodology as was used for the Authroize.Net credit card processing (i.e., it should utilize the API and not forms for the processing).

I've quickly reviewed the API and I believe that we need to implement their 'Vault' capability as well, which is the mechanism that allows us to retrieve a data key related to a customer's credit card and use that to process a payment, without us ever needing to store any credit card data in SA Billing.

Please take a look at this and give us an estimate on the development time.

Thanks,
Tim

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jul 4, 2017, 06:35

For details about the progress and effort, please check the following link:

#### [Effort Log for #305 ](http://gitlab.aavaz.biz/AnswerNet/SABilling/wikis/effort-log-305)

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jul 6, 2017, 08:09

@kdjstudios 

spent time on this to add moneris as cc payment processor for SCC. It might take 1-2 days (7 - 16 hr) in order to complete this.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jul 13, 2017, 03:10

@kdjstudios @tim.traylor   
In the initial SCC data dump there was a credit card data file, but that file was formatted we need a csv file.  
Means, ![Screenshot-5](/uploads/cb3ff18d4ffbcc5988b1e9a4832a697c/Screenshot-5.png) the file is in csv format and we cannot get the real numbers from that formatted cell.

**Need an updated unformatted csv file for sending bulk info on Pyment Vault**

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 13, 2017, 14:23

@smasih @tim.traylor 

I have attached two files, both removed formating. Please be cautious when opening a files in Excel. As by default, Microsoft Excel displays cells that contain a number with more than 12 characters as scientific notation. This does not mean that they are formated in the file.

I have also removed all spaces and dashes from the CC numbers column by using the find and replace tool in Excel. 

[creditcarddata-04_20_2017.csv](/uploads/be7343991a9ff6307e9899b3b78f6b96/creditcarddata-04_20_2017.csv)

[creditcarddata-04_20_2017.xlsx](/uploads/cf76563339c1e52ce3b11d685c2e27c0/creditcarddata-04_20_2017.xlsx)

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jul 18, 2017, 06:05

@kdjstudios 

Awesome, thanks for the files. 

We have added Moneris as well you may test using card   
* VISA: 4242424242424242
*MASTERCARD: 5454545454545454
*EXP: 0909

You can test the payments, adding the card on vault, deleting the card from vault updating the card from vault.

**Before you make any payment you need to add the card on file/vault so that the details are stored on moneris's server and can be retrieved while making the payment**

Please use dev.sabilling and Keener Comm. site for testing (I have updated the sandbox ip keys on site level)

Add on file button is on Account level.

please let us know if there was any issue.

thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 18, 2017, 08:24

@smasih 

Thank you I will be attempting to test now.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 18, 2017, 08:36

@smasih @tim.traylor 

Could you please explain "add the card to the vault" as I have no reference to a vault on SAB. If this is something that needs to happen on Moneris, do we have a test/sandbox account? If so what are the credentials. 

Steps attempted:
Logged into DEV
Located account (KNR-3032)
Clicked pay with CC
Input amount and click pay
Was presented with error "something went wrong" - I checked my email, but did not see any emails with errors sent out; presumably because this was on DEV.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jul 19, 2017, 04:33

@kdjstudios @tim.traylor 


1.  Let's pick any new site say, Towne.
2. Go to [admin](https://dev.sabilling.com/admin/security_levels) and click on **Credit-Card Gate Way**
3. Select Towne under Moneris.
4. Now go to [site edit page of Towne](https://dev.sabilling.com/sites/542363b697a7370b05000003/edit)
5. Under **Credit Card Fulfillment**  section update api keys (*we have sandbox keys at the moment*)
  * Api Login: **moneris** 
  * Api Key: **hurgle**
  * **Due to some error you won't be able to update the site we are trying still trying to figure out the error.** 
6. Now Jump to any Account of town and click on **Add Credit Card On File**
7. Fill in the details
  * Name on Card eg: John Doe
  * Card #: **5454545454545454 _/_ 4242424242424242**
  * Expiry: **0909**
  * Type: 54 is Master Card 42 is Visa
8. You will receive a confirmation, which means your card is added to the vault and you can make payments without adding details.
9. Go Back to main Account PAge and click on **Make Credit Card Payment**
10.   Enter amount in decimal (eg: 1.00) select account
11. You will receive a confirmation message along with the receipt.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 19, 2017, 09:13

@smasih 

After performing the steps listed above I received the following errors when performing on the Testy Site:


*  Step 5: "Something went wrong" this happened when trying to save the account; how ever it information appears to save. 


*  Step 7 & 8: "There was an error while UPDATING the card on Moneris Vault. ERROR: Can not find previous" this happened when trying to save the card information. 


* Step 10 & 11:  
"Can not find previous
Download receipt
Account number: TEST-00001
Cc last 4: 
Auth code: null
Transaction id: null
Card holder: 
Card type: 
Invoice: 
Payment media: 
Response code: 983
Amount: $1.00
Total amount charged: TRANSACTION FAILED
Trans date: (null) null
Resultcode: Can not find previous" This error occured when trying to submit the payment.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jul 20, 2017, 01:52

@kdjstudios 

there should be some configuration issue on the testy site, can we use Towne for testing on dev instead of Testy site. As we are on dev so any change in data would hardly matter. So please try with towne or keener or keener cali thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 20, 2017, 08:39

@smasih 

I just followed the below steps for account 0125 on Towne. The following issues I found: 

After updating/adding the CC card on file. there is no link back to the account.  The account name in the navigation at the top of the page is not click able.

Other then that the payment went through.  Marking as in development until all of the issues found are addressed.
